mkt/extensions/models.py

```diff
@@ -179,26 +120,165 @@ def mini_manifest(self):
+        try:
+            version = self.latest_public_version
+        except ExtensionVersion.DoesNotExist:
+            return {}
         mini_manifest = {
-            'name': self.manifest['name'],
-            'package_path': self.download_url,
-            'version': self.manifest['version']
+            'name': version.manifest['name'],
+            'package_path': version.download_url,
+            'version': version.manifest['version']
         }
-        if 'author' in self.manifest:
+        if 'author' in version.manifest:
             mini_manifest['developer'] = {
-                'name': self.manifest['author']
+                'name': version.manifest['author']
             }
-        if 'description' in self.manifest:
-            mini_manifest['description'] = self.manifest['description']
+        if 'description' in version.manifest:
+            mini_manifest['description'] = version.manifest['description']
         return mini_manifest
 
+    @property
+    def mini_manifest_url(self):
+        return absolutify(reverse('extension.mini_manifest',
+                                  kwargs={'uuid': self.uuid}))
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.clean_slug()
+        return super(Extension, self).save(*args, **kwargs)
+
+    def __unicode__(self):
+        return u'%s: %s' % (self.pk, self.name)
+
+    def update_status_according_to_versions(self):
+        """Update `status`, `latest_version` and `latest_public_version`
+        properties depending on the `status` on the ExtensionVersion
+        instances attached to this Extension."""
+        # If there is a public version available, the extension should be
+        # public. If not, and if there is a pending version available, it
+        # should be pending. Otherwise it should just be incomplete.
+        if self.versions.filter(status=STATUS_PUBLIC).exists():
+            self.update(status=STATUS_PUBLIC)
+        elif self.versions.filter(status=STATUS_PENDING).exists():
+            self.update(status=STATUS_PENDING)
+        else:
+            self.update(status=STATUS_NULL)
+        # Delete latest_version and latest_public_version properties, since
+        # they are writable cached_properties they will be reset the next time
+        # they are accessed.
+        try:
+            if self.latest_version:
+                del self.latest_version
+        except ExtensionVersion.DoesNotExist:
+            pass
+        try:
+            if self.latest_public_version:
+                del self.latest_public_version
+        except ExtensionVersion.DoesNotExist:
+            pass
+
+
+class ExtensionVersion(ModelBase):
+    extension = models.ForeignKey(Extension, related_name='versions')
+
+    default_language = models.CharField(default=settings.LANGUAGE_CODE,
+                                        max_length=10)
+    manifest = JSONField()
+    version = models.CharField(max_length=255, default='')
```

Comment:
Technically, version can only be 23 length, fwiw.  (x.x.x.x where x is an integer from 0 to 65535)